### ISSUE TITLE
Add a helper to determine if `root` should be excluded when considering `referenceRoots`

### ISSUE THREAD

**Issue by romainmenke** [stylelint/stylelint#9297]
With `referenceFiles` and `loader` it is possible that one of `result.stylelint.referenceRoots` is a bundle that already includes the current `root`.

The position in the AST of that embedded `root` will be more representative in terms of source order, conditions (e.g. `@media`, ...) and should be used instead of also considering the current `root` as passed to the rule directly.

Maybe we can add a helper to check if the current `root` is already embedded in one of the `referenceRoots` using `node.source.input.from`.

We should consider that `from` can be unknown and that the absolute paths must be equal.

_Originally posted by @romainmenke in https://github.com/stylelint/stylelint/pull/9292#discussion_r3252534444_

**Comment by romainmenke** [comment]
In the future we might also want to take cascade layers into account.
An example of this can be found in the PostCSS plugins for custom media, properties, ... https://github.com/csstools/postcss-plugins/blob/main/plugins/postcss-custom-media/src/cascade-layers.ts

**Comment by jeddy3** [comment]
> Maybe we can add a helper to check if the current `root` is already embedded in one of the `referenceRoots` using `node.source.input.from`.

Sounds good.

**Comment by Pedro-Muller29** [comment]
> Maybe `node.source.input.from` can be used?

Want to align on scope before opening a PR.

The naive helper walks each `referenceRoot` once per call, and the call happens once per rule per linted file. For a project with N files linted against the same bundle, that's N × R walks of the same bundle (R = rules using the helper).

The shape that captures the real win is one walk per bundle, reused across files. That implies the cache lives above `lintSource`, on the `stylelint` instance, or hanging off `config._resolvedReferenceFiles`, since `getReferenceRoots` reparses the bundle on every `lintSource` call today.

Is the intended scope just the helper (no cache, accept the re-walks), or does it include the cross-file caching? Anything I'm reading wrong on the flow?

**Comment by romainmenke** [comment]
I think we can cache on the module and use `WeakMap`'s.

My current best attempt:

```js
import { isBoolean } from './validateTypes.mjs';

/** @import { Root, Node } from 'postcss' */

/** @type {WeakMap<Root, WeakMap<Node, boolean>>} */
const cache = new WeakMap();

/**
 * Check whether a root is embedded in some reference roots.
 *
 * @param {Root} root
 * @param {Root[]} referenceRoots
 * @returns {boolean}
 */
export default function isRootEmbeddedInSomeReferenceRoots(root, referenceRoots) {
	const from = getFrom(root);

	if (!from) return false;

	for (const referenceRoot of referenceRoots) {
		const cachedLookup = cache.get(referenceRoot) ?? new WeakMap();

		let isIncluded = cachedLookup.get(root);

		if (isIncluded) return true;

		if (!isBoolean(isIncluded)) {
			isIncluded = false;

			referenceRoot.walk((childNode) => {
				if (getFrom(childNode) !== from) return;

				isIncluded = true;

				return false;
			});

			cachedLookup.set(root, isIncluded);
			cache.set(referenceRoot, cachedLookup);
		}

		if (isIncluded) return true;
	}

	return false;
}

/**
 * @param {Node} node
 * @returns {string | undefined}
 */
function getFrom(node) {
	const from = node.source?.input.from;

	if (!from) return;

	// When parsing with postcss without `from` `node.source?.input.from` has a value that looks like `<input css w_2MOM>`
	if (from.startsWith('<input css')) return;

	return from;
}
```

```js
import isRootEmbeddedInSomeReferenceRoots from '../isRootEmbeddedInSomeReferenceRoots.mjs';
import postcss from 'postcss';


describe('isRootEmbeddedInSomeReferenceRoots', () => {
	it('is true for the same root', () => {
		const rootA = postcss.parse('.foo {}', { from: 'foo.css' });
		const rootB = postcss.parse('.foo {}', { from: 'foo.css' });

		expect(isRootEmbeddedInSomeReferenceRoots(rootA, [rootB])).toBe(true);

		// internal caches must not affect the outcome
		expect(isRootEmbeddedInSomeReferenceRoots(rootA, [rootB])).toBe(true);
	});

	it('is false for an unrelated root', () => {
		const rootA = postcss.parse('.foo {}', { from: 'foo.css' });
		const rootB = postcss.parse('.bar {}', { from: 'bar.css' });

		expect(isRootEmbeddedInSomeReferenceRoots(rootA, [rootB])).toBe(false);

		// internal caches must not affect the outcome
		expect(isRootEmbeddedInSomeReferenceRoots(rootA, [rootB])).toBe(false);
	});

	it('is false for roots without `from`', () => {
		const rootA = postcss.parse('.foo {}');
		const rootB = postcss.parse('.foo {}');

		expect(isRootEmbeddedInSomeReferenceRoots(rootA, [rootB])).toBe(false);
	});

	it('is true for an embedded root', () => {
		const rootA = postcss.parse('.foo {}', { from: 'foo.css' });

		const rootB = postcss.parse('.foo {}', { from: 'foo.css' });
		const rootC = postcss.parse('.bar {}', { from: 'bar.css' });

		rootC.first.append(...rootB.nodes);

		expect(isRootEmbeddedInSomeReferenceRoots(rootA, [rootC])).toBe(true);

		// internal caches must not affect the outcome
		expect(isRootEmbeddedInSomeReferenceRoots(rootA, [rootC])).toBe(true);
	});
});
```

----

I am still a bit concerned about the load of this helper.

Especially when the reference roots aren't bundles at all.
If one of the reference roots is a bundle and happens to have a node early in the tree, then the walking is quickly aborted.

But when they aren't bundles, then we walk these AST's for no reason at all.

Even with aggressive caching it feels a bit wasteful.

Maybe we can use the absence of a `loader` as signal to not do this lookup?

**Comment by romainmenke** [comment]
🤔 We can also first do a simple iteration of top level nodes and only then do a full walk if nothing was found.
This should find a match in most cases and is a lot quicker/cheaper.

**Comment by jeddy3** [comment]
> We can also first do a simple iteration of top level nodes and only then do a full walk if nothing was found.

SGTM. The feature is marked as experimental, so we're free to try out these approaches and see what sticks.
